---
title: "YF old young analysis"
author: "Mikhail Shugay, Anna Obraztsova"
date: "15/11/2016"
output: pdf_document
---

Load data. Data was generated by running MIGMAP/Analyze, see [here](https://github.com/mikessh/migmap).

```{r, warning=FALSE, message=FALSE}
old <- c("Abdulain", "Ilgen", "Mamaev", "Smirnov", "Vlasov", "Xwoman")
young <- c("Antipyat1", "Antipyat2", "Epifancev1", "Epifancev2", "Hadjibekov", "Koshkin", "Kovalchuk")

df <- data.frame()

for (sample in old) {
  .df <- read.table(paste('out2/yf_old_DNA/', sample, ".shm.txt", sep = ""), header=T, sep="\t")
	.df$proj <- "old"
	.df$sample <- sample
	df <- rbind(df, .df)
}

for (sample in young) {
	.df <- read.table(paste('out2/yf_young_DNA/', sample, ".shm.txt", sep = ""), header=T, sep="\t")
	.df$proj <- "young"
	.df$sample <- sample
	df <- rbind(df, .df)
}

```

Summarize by ``V/J + mutation`` entries and remove alleles.

```{r, warning=FALSE, message=FALSE}
library(plyr)
df.1 <- ddply(df, .(proj, sample, segment, segment.name, region, mutation.type, pos.nt, from.nt, to.nt, pos.aa, from.aa, to.aa),
              summarize, rate = sum(freq), total.clonotypes = sum(clonotypes), 
              allele.rate = sum(freq)/segment.freq[1], allele.clonotype.share = sum(clonotypes) / segment.clonotypes[1])

df.2 <- subset(df.1, total.clonotypes <= 5 || allele.rate < 0.45 || allele.clonotype.share < 0.45)
```

Mutation density on IG sequence. Mutations are split in the following 4 categories:

- S, silent
- R, replacement
- LOF, loss of function. Either substitution leading to stop codon or out-of-frame indel
- Indel, in-frame insertion or deletion

```{r, warning=FALSE, message=FALSE}
library(ggplot2)

df.2$from.nt <- as.character(df.2$from.nt)
df.2$to.nt <- as.character(df.2$to.nt)
df.2$from.aa <- as.character(df.2$from.aa)
df.2$to.aa <- as.character(df.2$to.aa)

df.2$type <- with(df.2, ifelse(mutation.type == "Substitution", ifelse(to.aa == "*", "Stop", ifelse(from.aa != to.aa, "R", "S")),
                               ifelse(nchar(from.nt) + nchar(to.nt) %% 3 != 0, "Shift", "Indel")))
df.2$region <- factor(df.2$region, c("FR1","CDR1","FR2","CDR2","FR3","CDR3","FR4"))

# Split by region

ggplot(df.2, aes(x=pos.aa, weight=rate, color=type)) + geom_density() + 
  facet_grid(proj~region, scales="free") + theme_bw()

# Full length

ggplot(df.2, aes(x=pos.aa, weight=rate, color=type)) + geom_density() + 
  facet_grid(~proj) + theme_bw()
```

Boxplots comparing share of each mutation type, i.e. its fraction across all errors in a given sample and IG region. Weighted by clonotype frequency:

```{r, warning=FALSE, message=FALSE, fig.height=8}
library(dplyr)

df.3 <- ddply(df.2, .(proj, sample, region), summarize,
              share.S = sum(rate[which(type == "S")]) / sum(rate),
              share.R = sum(rate[which(type == "R")]) / sum(rate),
              logRS = log(sum(rate[which(type == "R")]) / sum(rate[which(type == "S")])),
              share.Shift = sum(rate[which(type == "Shift")]) / sum(rate),
              share.Indel = sum(rate[which(type == "Indel")]) / sum(rate))

library(reshape2)

df.4 <- melt(df.3, id.vars = c("proj", "sample", "region"))

ggplot(df.4, aes(x=proj, group=proj, y=value)) + geom_boxplot() + facet_grid(variable~region, scales="free") + theme_bw()
```

RS ratio separatly

```{r, warning=FALSE, message=FALSE, fig.height=6}
df.3 <- ddply(df.2, .(proj, sample, region), summarize,
              RS = sum(rate[which(type == "R")]) / sum(rate[which(type == "S")]))
df.4 <- melt(df.3, id.vars = c("proj", "sample", "region"))
ggplot(df.4, aes(x=proj, group=proj, y=value)) + geom_boxplot() + facet_grid(variable~region, scales="free") + theme_bw()
```

Unweighted:

```{r, warning=FALSE, message=FALSE, fig.height=6}
df.3 <- ddply(df.2, .(proj, sample, region), summarize,
              share.S = sum(total.clonotypes[which(type == "S")]) / sum(total.clonotypes),
              share.R = sum(total.clonotypes[which(type == "R")]) / sum(total.clonotypes),
              share.Shift = sum(total.clonotypes[which(type == "Shift")]) / sum(total.clonotypes),
              share.Indel = sum(total.clonotypes[which(type == "Indel")]) / sum(total.clonotypes))

library(reshape2)
df.4 <- melt(df.3, id.vars = c("proj", "sample", "region"))
ggplot(df.4, aes(x=proj, group=proj, y=value)) + geom_boxplot() + facet_grid(variable~region, scales="free") + theme_bw()
```

Length distribution of indels

```{r, warning=FALSE, message=FALSE, fig.show='hold'}
ggplot(filter(df.2, mutation.type == 'Insertion'), aes(x=nchar(to.nt), fill=proj)) + geom_histogram(position='dodge', binwidth=1) + theme_bw() + xlab('Insertion length')
ggplot(filter(df.2, mutation.type == 'Deletion'), aes(x=nchar(from.nt), fill=proj)) + geom_histogram(position='dodge', binwidth=1) + theme_bw() + xlab('Deletion length')
```

Substitution pattern
```{r, warning=FALSE, message=FALSE}
library(seqLogo)
library(Biostrings)

pfmO <- consensusMatrix(readDNAStringSet('out2/sites/old_s.fasta'))
pfmY <- consensusMatrix(readDNAStringSet('out2/sites/young_s.fasta'))
pfmO <- as.data.frame(t(pfmO[1:4,]))
pfmY <- as.data.frame(t(pfmY[1:4,]))

proportion <- function(x){
   rs <- sum(x);
   return(x / rs);
}

pwmO <- apply(pfmO, 1, proportion)
pwmO <- makePWM(pwmO)
seqLogo(pwmO, yaxis=FALSE)

pwmY <- apply(pfmY, 1, proportion)
pwmY <- makePWM(pwmY)
seqLogo(pwmY, yaxis=FALSE)
```

Isotype distribution
```{r, warning=FALSE, message=FALSE}
rna <- data.frame()

old_rna = c("Abdulain", "Ilgen", "Mamaev", "Smirnov", "Vlasov")
young_rna = c("Antipyat", "Epifancev", "Hadjibekov", "Koshkin", "Kovalchuk")

for (sample in old_rna){
  .df <- read.table(paste('data16.11/yf_old_RNA/', sample, ".txt", sep = ""), header=T, sep="\t")
	.df$proj <- "old"
	.df$sample <- sample
	rna <- rbind(rna, .df)
}

for (sample in young_rna){
	.df <- read.table(paste('data16.11/yf_young_RNA/', sample, ".txt", sep = ""), header=T, sep="\t")
	.df$proj <- "young"
	.df$sample <- sample
	rna <- rbind(rna, .df)
}

new_colnames = c('clone.id','clone.count','clone.fraction','clonal.seq','clonal.seq.qual','all.v.hits','all.d.hits',
	'all.j.hits','all.c.hits','all.v.alignments','all.d.alignments','all.j.alignments','all.c.alignments',
	'nt.seq.FR1','min.qual.FR1','nt.seq.CDR1','min.qual.CDR1','nt.seq.FR2','min.qual.FR2','nt.seq.CDR2','min.qual.CDR2',
	'nt.seq.FR3','min.qual.FR3','nt.seq.CDR3','min.qual.CDR3','nt.seq.FR4','min.qual.FR4','aa.seq.FR1','aa.seq.CDR1',
	'aa.seq.FR2','aa.seq.CDR2','aa.seq.FR3','aa.seq.CDR3','aa.seq.FR4','ref.points', 'proj', 'sample')

colnames(rna) <- new_colnames
rna <- mutate(rna, isotype = str_sub(all.c.hits, 1, 4))

rna.2 <- ddply(rna, .(proj, sample), summarize,
             share.IGHA = sum(clone.fraction[which(isotype == "IGHA")]),
             share.IGHD = sum(clone.fraction[which(isotype == "IGHD")]),
             share.IGHE = sum(clone.fraction[which(isotype == "IGHE")]),
             share.IGHG = sum(clone.fraction[which(isotype == "IGHG")]),
             share.IGHM = sum(clone.fraction[which(isotype == "IGHM")]))

rna.3 <- melt(rna.2, id.vars = c("proj", "sample"))
ggplot(rna.3, aes(x=proj, group=proj, y=value)) + geom_boxplot() + facet_grid(.~variable, scales="free") + theme_bw()
```

